src/java.base/share/classes/java/util/zip/ZipFile.java

```diff
@@ -1389,34 +1422,49 @@ private static boolean isZip64ExtBlockSizeValid(int blockSize, long csize,
         private int tablelen;                // number of hash heads
 
         /**
-         * A class representing a key to a ZIP file. A key is based
-         * on the file key if available, or the path value if the
-         * file key is not available. The key is also based on the
-         * file's last modified time to allow for cases where a ZIP
-         * file is re-opened after it has been modified.
+         * A class representing a key to the Source of a ZipFile.
+         * The Key is composed of:
+         * - The BasicFileAttributes.fileKey() if available, or the Path of the ZIP file
+         * if the fileKey() is not available.
+         * - The ZIP file's last modified time (to allow for cases
+         * where a ZIP file is re-opened after it has been modified).
+         * - The Charset, that was provided when constructing a ZipFile instance,
+         * for reading non-UTF-8 entry names and comments.
+         * The unique combination of these components identifies a Source of a ZipFile.
          */
         private static class Key {
-            final BasicFileAttributes attrs;
-            File file;
-            final boolean utf8;
-
-            public Key(File file, BasicFileAttributes attrs, ZipCoder zc) {
+            private final BasicFileAttributes attrs;
+            private final File file;
+            // the Charset to be used for processing non-UTF-8 entry names in the ZIP file
```

Comment:
Similarly this could just say "The Charset that was provided when constructing the ZipFile instance"

Reply:
Yes, I think your suggested comment is good. I've updated the PR to use this.